core/modules/replica/Job.cc

```diff
@@ -207,6 +213,19 @@ void Job::start() {
 }
 
 
+void Job::wait() {
+ 
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__);
+
+    if (state() == State::FINISHED) return;
+
+    unique_lock<mutex> lock(_onFinishMtx);
+    _onFinishCv.wait(lock, [this] {
+        return state() == State::FINISHED;
+    });
```

Comment:
I think this has a chance to block forever. I'll approve PR if you explain how it can happen (and fix it) 😉 

Also condition variables are messy in general, maybe `std::future` would be easier to use for this kind of pattern?

Reply:
I've added `unique_lock<mutex> lock(_onFinishMtx);` before modifying `_state`. Now, both update and acquisition of the state are mutually exclusive for the purpose of implementing the proposed synchronization. Otherwise, there was a chance of missing the `notify_all` signal.
Regarding `std::future`, I find it messy and counter-intuitive.

Reply:
Fix looks fine.

And you need to work in you intuition, `std::future` is supposed to be more intuitive than condition variables :) I think your use case can be implemented trivially with something like this:

    class Job {

        std::promise<void> _finishPromise;
        std::future<void> _finishFuture;

        Job() : _finishFuture(_finishPromise.get_future()) {}

        void finish() {
            // Unblock threads (if any) waiting on the synchronization...
            _finishPromise.set_value();
        }

        void wait() { _finishFuture.wait(); }

    };